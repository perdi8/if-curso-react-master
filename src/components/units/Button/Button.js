import {FontAwesomeIcon} from '@fortawesome/react-fontawesome';
import React from 'react';

// Styles
import {ButtonStyled} from './Button.styles.js';

// Como evitar el error de onClick en la prop.
// 1. Dejando un valor por defecto onClick = () => {}
// 2. Testear todas las props (muy costoso en tiempo y mantenimiento)
// 3. Utilizar typescript.
// Typescript/Flow reduce notablemente la necesidad de realizar test unitarios sobre nuestros componentes.

function Button({
	text,
	type,
	onClick,
	isDisabled = false,
	icon,
	iconPosition,
	textStyles,
	iconStyles,
	buttonStyles,
}) {
	const handleClick = () => {
		if (isDisabled) {
			return;
		}
		if (!onClick) {
			return;
		}
		onClick();
	};

	return (
		<ButtonStyled
			type={type}
			onClick={handleClick}
			disabled={isDisabled}
			style={{...buttonStyles}}
		>
			{iconPosition === 'left' && icon ? (
				<FontAwesomeIcon icon={icon} style={{...iconStyles}} />
			) : null}
			<span style={{...textStyles}}>{text}</span>
			{iconPosition === 'right' && icon ? <FontAwesomeIcon icon={icon} /> : null}
		</ButtonStyled>
	);
}

export default Button;
