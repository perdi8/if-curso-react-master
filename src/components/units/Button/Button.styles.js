import styled from 'styled-components';
import Colors from 'theme/Colors';

export const ButtonStyled = styled.button`
	color: ${Colors.mainColor};
	background-color: ${Colors.secondaryColor};
	border-radius: 6px;
	padding: 0.5rem 1.75rem;
	border: 1px solid #ccc;

	&:hover {
		cursor: pointer;
	}
`;
