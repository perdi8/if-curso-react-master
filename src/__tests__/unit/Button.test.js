import React from 'react';
import {fireEvent, render} from '@testing-library/react';
import Button from 'components/units/Button/Button';
import renderer from 'react-test-renderer';

describe('Button component', () => {
	it('renders correctly', () => {
		const {getByText} = render(<Button text="Hola" />);
		expect(getByText('Hola')).toBeInTheDocument();
	});

	it('has proper snapshot', () => {
		const ButtonComponent = renderer.create(<Button text="Hola" />);
		let tree = ButtonComponent.toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('handles Click event', () => {
		const handleClick = jest.fn();
		const {getByText} = render(<Button text="Clicked" onClick={handleClick} />);
		fireEvent.click(getByText(/Clicked/i));
		expect(handleClick).toHaveBeenCalledTimes(1);
	});
});
