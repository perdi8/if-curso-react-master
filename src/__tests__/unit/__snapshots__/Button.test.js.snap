// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component has proper snapshot 1`] = `
<button
  className="sc-bdnxRM hNJinF"
  disabled={false}
  onClick={[Function]}
  style={Object {}}
>
  <span
    style={Object {}}
  >
    Hola
  </span>
</button>
`;
